Subshell issue when building in a Python 3 environment

leandro-lucarella-sociomantic opened the pull request:
This error happens when building when `python` is Python 3:

```
$ make install
sed 's/^:Version: devel$/:Version: 0.11.1-4-g2b1f532-dirty/' man.rst | \
   rst2man --exit-status=1 > git-hub.1 || (rm -f git-hub.1 && false)
./generate-bash-completion git-hub > bash-completion || (rm -f bash-completion && false)                                Traceback (most recent call last):
File "git-hub", line 46, in <module>
import urllib2
ModuleNotFoundError: No module named 'urllib2'
make: *** [Makefile:28: bash-completion] Error 1
```

The `generate-bash-completion` script has `PYTHON=${PYTHON:-python}`. That fixes the issue if I `export PYTHON=python2` and run the script directly, but when `make` runs the script, `PYTHON` is not defined and the above error still occurs.

leandro-lucarella-sociomantic commented:
Ugh? Is there any system out there that defines `python` as `python3`? :scream: 

In any case, I guess at some point we should just convert all to python3.

mihails-strasuns-sociomantic commented:
> Ugh? Is there any system out there that defines python as python3

Of course, and there will be more of them over time.

jean commented:
On [Termux](https://termux.com/), `python` is 3.6.2

leandro-lucarella-sociomantic commented:
And no python2 available? Even if using `python2` as the interpreter?

leandro-lucarella-sociomantic committed:
Make python2 the default interpreter

Some installations have python 3 as the `python` binary, which breaks
the generate completion script.

Fixes #223.

leandro-lucarella-sociomantic commented:
Fix attached. Can you verify if it works?

leandro-lucarella-sociomantic commented:
ping @jean

jean commented:
Sorry, I was being dozy. My `export PYTHON=python2` wasn't working because the makefile overrides `PYTHON`. So the real fix is to use the `PYTHON` from the environment if it exists:

```
diff --git a/Makefile b/Makefile
index be270e0..ba0e247 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 
 prefix ?= /usr/local
 
-export PYTHON := python
+export PYTHON ?= python
```

If I make that change, then `generate-bash-completion` works.

jean reviewed (commented):
This wouldn't work for running via make, because `export PYTHON := python` in `Makefile` ensures that `PYTHON` will be defined here, so `python2` will not be used. 

Change to `export PYTHON ?= python` in the `Makefile`, then `PYTHON=python2 make install` will work.

Or change to `export PYTHON ?= python2` in the `Makefile`, then plain `make install` will work on systems where `python` is Python 3, while still allowing override on the commandline (like `PYTHON=python3 make install`).

leandro-lucarella-sociomantic commented:
> My export PYTHON=python2 wasn't working because the makefile overrides PYTHON. So the real fix is to use the PYTHON from the environment if it exists:

To override the make variable just use `make PYTHON=python2`.

But if `python` is being shipped as `python3` the safest approach is to use `python2` explicitly I guess, I would commit this fix if it does work for you.

leandro-lucarella-sociomantic commented:
I just pushed a few more fixes to set `python2` as the default.

mihails-strasuns-sociomantic commented:
> the safest approach is to use python2 explicitly I guess

I approve :)

AFAIK it is possible even in Debian to configure python3 to be the default via update-alternatives (though plenty of scripts will break because they assume python2)

leandro-lucarella-sociomantic committed:
make: Allow overriding PYTHON via env vars

The `PYTHON` variable can be overridden only via the `make PYTHON=...`
syntax. This commit makes it possible to override it via environment
variables too.

leandro-lucarella-sociomantic commented:
> Change to export PYTHON ?= python

I added a commit to do this too, but I'm not sure if this is something I want to be picked up from the environment. What the `PYTHON` variable is here, is just a configuration variable, not something to set in the environment to persist. I guess in practice it will be OK, but it makes some noise in my head.

leandro-lucarella-sociomantic commented:
> AFAIK it is possible even in Debian to configure python3 to be the default via update-alternatives (though plenty of scripts will break because they assume python2)

Nope, I think this was left out of the debian alternatives system just because of this. It would break too much stuff to change the default python interpreter.

Python itself recommends leaving `python` pointing to `python2` *for the time being*: http://legacy.python.org/dev/peps/pep-0394/

Arch Linux mess was the one that triggered the creation of a PEP for this :rofl:

mihails-strasuns-sociomantic commented:
Well, it also says..

> in preparation for an eventual change in the default version of Python, Python 2 only scripts should either be updated to be source compatible with Python 3 or else to use python2 in the shebang line

.. which is exactly what is being suggested here ;)

leandro-lucarella-sociomantic commented:
@jean can you give a final look/test so I can merge this?

jean commented:
Will do tomorrow morning, sorry I'm slow ..

On 21 Sep 2017 9:01 p.m., "Leandro Lucarella" <notifications@github.com>
wrote:

> @jean <https://github.com/jean> can you give a final look/test so I can
> merge this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sociomantic-tsunami/git-hub/pull/223#issuecomment-331165646>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFLQGkYu8Hf1jCZ9msDRwVCZTAheAUsks5skmw4gaJpZM4PSPpn>
> .
>

leandro-lucarella-sociomantic commented:
> Will do tomorrow morning, sorry I'm slow ..

Sure! Thanks!

jean reviewed (commented):
Works for me 👍
